src/Streamly/Prelude.hs

```diff
@@ -120,6 +123,20 @@ unfoldrM step = fromStream . go
 each :: (Streaming t, Foldable f) => f a -> t m a
 each = Prelude.foldr cons nil
 
+-- | Iterate a pure function from a seed value, streaming the results forever
+iterate :: Streaming t => (a -> a) -> a -> t m a
+iterate step = fromStream . go
+  where
+  go s = scons s (Just (go (step s)))
+
+-- | Iterate a monadic function from a seed value, streaming the results forever
+iterateM :: (Streaming t, Monad m) => (a -> m a) -> m a -> t m a
```

Comment:
The seed must be pure, and not monadic. Only the step should be monadic. See `unfoldrM` which is very similar to `iterateM`.

Reply:
I see that `streaming` is using a monadic seed. But I am not sure if there is a clear advantage in doing that. We can always run the action separately and pass the seed to this function. So a pure seed should be enough to achieve all that a monadic seed can. Pure seed is simpler. I would go with pure unless there is some clear benefit proven that impure seed gives us.

Reply:
In the other monadic fold functions in the prelude `streamly` uses a monadic seed. Is there any particular advantage to that in folds?

Reply:
The general rule is: inputs are pure and outputs are monadic, remember the `a -> m b` in bind. That's what has been chosen as convention for convenient composition. In `unfoldr` and `iterate` it is a `seed` i.e. an input, in folds it is an `accumulator` i.e. an output, that's why it is monadic. We cannot make it pure without making the function itself pure.